core/market/tests/test_app_session_id.rs

```diff
@@ -0,0 +1,549 @@
+use all_asserts::*;
+use anyhow::Result;
+use chrono::{Duration, Utc};
+
+use ya_market::testing::agreement_utils::negotiate_agreement;
+use ya_market::testing::proposal_util::exchange_proposals_exclusive;
+use ya_market::testing::MarketsNetwork;
+use ya_market::testing::OwnerType;
+
+use ya_client::model::market::AgreementOperationEvent as AgreementEvent;
+
+const REQ_NAME: &str = "Node-1";
+const PROV_NAME: &str = "Node-2";
+
+/// Create Agreements with different session ids. Query Agreement
+/// events using different session ids on both Provider and Requestor side.
+#[cfg_attr(not(feature = "test-suite"), ignore)]
+#[actix_rt::test]
+#[serial_test::serial]
+async fn test_session_events_filtering() -> Result<()> {
+    let network = MarketsNetwork::new(None)
+        .await
+        .add_market_instance(REQ_NAME)
+        .await?
+        .add_market_instance(PROV_NAME)
+        .await?;
+
+    let req_market = network.get_market(REQ_NAME);
+    let req_engine = &req_market.requestor_engine;
+    let req_id = network.get_default_id(REQ_NAME);
+    let prov_id = network.get_default_id(PROV_NAME);
+    let prov_market = network.get_market(PROV_NAME);
+
+    let num = 4;
+
+    let mut proposals = Vec::new();
+    for i in 0..num {
+        proposals.push(
+            exchange_proposals_exclusive(&network, REQ_NAME, PROV_NAME, &format!("neg-{}", i))
+                .await
+                .unwrap()
+                .proposal_id,
+        );
+    }
+
+    let mut agreements = vec![];
+    for proposal_id in proposals.iter() {
+        let agreement_id = req_engine
+            .create_agreement(
+                req_id.clone(),
+                &proposal_id,
+                Utc::now() + Duration::hours(1),
+            )
+            .await
+            .unwrap();
+        agreements.push(agreement_id);
+    }
+
+    // Create two session names.
+    let mut sessions = agreements[..num - 1]
+        .iter()
+        .map(|_| "session-1".to_string())
+        .collect::<Vec<_>>();
+    sessions.push("session-2".to_string());
+
+    let confirm_timestamp = Utc::now();
+    for (agreement_id, session_id) in agreements.iter().zip(sessions.iter()) {
+        req_engine
+            .confirm_agreement(req_id.clone(), &agreement_id, Some(session_id.clone()))
+            .await
+            .unwrap();
+    }
+
+    let ids = agreements.clone();
+    let session_ids = sessions.clone();
+    let query_handle = tokio::task::spawn_local(async move {
+        for (agreement_id, session_id) in ids.iter().zip(session_ids.iter()) {
```

Comment:
It's not corner case at all. We do this on two separates nodes.

Reply:
yes, exactly. And thats why it is extremely improbable that both agents will use the same sessions being independent.

Reply:
Purpose of this test isn't checking what happens if session id is the same. I used same ids, because it doesn't matter and was easier to have one vectors of ids than 2